snapshotter/devmapper/pool_device.go

```diff
@@ -73,87 +73,190 @@ func NewPoolDevice(ctx context.Context, config *Config) (*PoolDevice, error) {
 	}, nil
 }
 
+// transition invokes 'fn' callback to perform devmapper operation and reflects device state changes/errors in meta store.
+// 'tryingState' will be set before invoking callback. If callback succeeded 'successState' will be set, otherwise
+// error details will be recorded in meta store.
+func (p *PoolDevice) transition(ctx context.Context, deviceName string, tryingState DeviceState, successState DeviceState, fn func() error) error {
```

Comment:
nit: I would probably make this a graph of states. Something like
```go
// or have this defined in the PoolDevice
var stateGraph = map[DeviceState]DeviceState{
    Resuming: Resumed,
    // and so on.
}

func (p *PoolDevice) transition(ctx context.Context, deviceName, state DeviceState, fn func() error) error {
}
```

Reply:
I'm not sure about this idea, this adds one more place to manage device states.
So if you forget to update keys, you'll get runtime error, which is not good.
Also I think it's better to explicitly define from/to state we're transitioning.

Reply:
I tend to leave it as is.